python_modules/dagster/dagster/_core/definitions/asset_graph.py

```diff
@@ -1,56 +1,122 @@
+                partition_mappings_by_key.update(
+                    {
+                        key: asset._partition_mappings  # pylint: disable=protected-access
+                        for key in asset.keys
+                    }
+                )
+                partitions_defs_by_key.update({key: asset.partitions_def for key in asset.keys})
+                group_names_by_key.update(asset.group_names_by_key)
             else:
                 check.failed(f"Expected SourceAsset or AssetsDefinition, got {type(asset)}")
+        return AssetGraph(
+            asset_dep_graph=generate_asset_dep_graph(assets_defs, source_assets),
+            source_asset_keys={source_asset.key for source_asset in source_assets},
+            partitions_defs_by_key=partitions_defs_by_key,
+            partition_mappings_by_key=partition_mappings_by_key,
+            group_names_by_key=group_names_by_key,
+        )
 
-        self.assets_defs = assets_defs
-        self.all_assets_by_key = {
-            asset_key: assets_def for assets_def in assets_defs for asset_key in assets_def.keys
-        }
-
-        self.asset_dep_graph = generate_asset_dep_graph(assets_defs, source_assets)
-        self.source_asset_keys = {source_asset.key for source_asset in source_assets}
+    @staticmethod
+    def from_external_assets(external_asset_nodes: Sequence["ExternalAssetNode"]) -> "AssetGraph":
+        upstream = {}
+        downstream = {}
+        partitions_defs_by_key = {}
+        group_names_by_key = {}
+
+        for node in external_asset_nodes:
+            upstream[node.asset_key] = {dep.upstream_asset_key for dep in node.dependencies}
+            downstream[node.asset_key] = {dep.downstream_asset_key for dep in node.depended_by}
+            partitions_defs_by_key[node.asset_key] = (
+                node.partitions_def_data.get_partitions_definition()
+                if node.partitions_def_data
+                else None
+            )
+            group_names_by_key[node.asset_key] = node.group_name
+
+        return AssetGraph(
+            asset_dep_graph={"upstream": upstream, "downstream": downstream},
+            source_asset_keys=set(),  # TODO
+            partitions_defs_by_key=partitions_defs_by_key,
+            partition_mappings_by_key=dict(),  # TODO
+            group_names_by_key=group_names_by_key,
+        )
 
     @property
     def all_asset_keys(self) -> AbstractSet[AssetKey]:
-        return self.all_assets_by_key.keys()
+        return self.asset_dep_graph["upstream"].keys()
+
+    def get_partitions_def(self, asset_key: AssetKey) -> Optional[PartitionsDefinition]:
+        return self.partitions_defs_by_key.get(asset_key)
+
+    def get_partition_mapping(
+        self, asset_key: AssetKey, in_asset_key: AssetKey
+    ) -> PartitionMapping:
+        partitions_def = self.get_partitions_def(asset_key)
+        partition_mappings = self.partition_mappings_by_key.get(asset_key) or {}
+        return partition_mappings.get(
```

Comment:
This is duplicating logic inside `AssetsDefinition.get_partition_mapping`, right? I think worth factoring it out into a common function so we don't accidentally change one without changing the other.